R/memoise.r

```diff
@@ -35,64 +35,132 @@
 #' @aliases memoise memoize
 #' @export memoise memoize
 #' @importFrom digest digest
 #' @examples
-#' # a() is evaluated anew each time. memA() is only re-evaluated 
+#' # a() is evaluated anew each time. memA() is only re-evaluated
 #' # when you call it with a new set of parameters.
 #' a <- function(n) { runif(n) }
 #' memA <- memoise(a)
 #' replicate(5, a(2))
 #' replicate(5, memA(2))
-#' 
+#'
 #' # Caching is done based on parameters' value, so same-name-but-
 #' # changed-value correctly produces two different outcomes...
 #' N <- 4; memA(N)
 #' N <- 5; memA(N)
-#' # ... and same-value-but-different-name correctly produces 
+#' # ... and same-value-but-different-name correctly produces
 #' #     the same cached outcome.
 #' N <- 4; memA(N)
 #' N2 <- 4; memA(N2)
-#' 
-#' # memoise() doesn't know about default parameters.
-#' memB <- memoise(function(n, dummy="a") { runif(n) })
+#'
+#' # memoise() knows about default parameters.
+#' b <- function(n, dummy="a") { runif(n) }
+#' memB <- memoise(b)
 #' memB(2)
 #' memB(2, dummy="a")
-#' # It doesn't know about parameter relevance, either. 
-#' # Different call means different cacheing, no matter 
+#' # This works, because the interface of the memoised function is the same as
+#' # that of the original function.
+#' formals(b)
+#' formals(memB)
+#' # However, it doesn't know about parameter relevance.
+#' # Different call means different cacheing, no matter
 #' # that the outcome is the same.
 #' memB(2, dummy="b")
-#' 
-#' # You can create multiple memoisations of the same function, 
-#' # and they'll be independent. 
+#'
+#' # You can create multiple memoisations of the same function,
+#' # and they'll be independent.
 #' memA(2)
 #' memA2 <- memoise(a)
 #' memA(2)  # Still the same outcome
 #' memA2(2) # Different cache, different outcome
-#' 
-#' # Don't do the same memoisation assignment twice: a brand-new 
-#' # memoised function also means a brand-new cache, and *that* 
+#'
+#' # Don't do the same memoisation assignment twice: a brand-new
+#' # memoised function also means a brand-new cache, and *that*
 #' # you could as easily and more legibly achieve using forget().
-#' # (If you're not sure whether you already memoised something,  
+#' # (If you're not sure whether you already memoised something,
 #' #  use is.memoised() to check.)
 #' memA(2)
 #' memA <- memoise(a)
 #' memA(2)
-memoise <- memoize <- function(f) {
+memoise <- memoize <- function(f, envir = parent.frame()) {
+  # We must not even try evaluating f -- once we start, there's no way back
+  if (inherits(try(eval.parent(substitute(f)), silent = TRUE), "try-error")) {
+    warning("Can't access f -- using old-style memoisation with dots interface. ",
+            "Define the memoised function before memoising to avoid this warning.")
+    memoise_old(f)
+  } else {
+    memoise_new(f, envir)
+  }
+}
+
+#' @importFrom stats setNames
+memoise_new <- function(f, envir) {
+  f_formals <- formals(f)
```

Comment:
I think you also need a test for primitive functions (e.g. `+`), which don't have formals. They should use `memoise_old()` without a warning.